x/ecocredit/migrations/v3/params.go

```diff
@@ -0,0 +1,182 @@
+)
+
+// TODO: Revisit this once we have proper gas fee framework.
+// Tracking issues https://github.com/cosmos/cosmos-sdk/issues/9054, https://github.com/cosmos/cosmos-sdk/discussions/9072
+const GasCostPerIteration = uint64(10)
+
+// TODO: remove after we open governance changes for precision
+const (
+	PRECISION uint32 = 6
+)
+
+// ParamKeyTable returns the parameter key table.
+func ParamKeyTable() paramtypes.KeyTable {
+	return paramtypes.NewKeyTable().RegisterParamSet(&Params{})
+}
+
+// Implements params.ParamSet
+func (p *Params) ParamSetPairs() paramtypes.ParamSetPairs {
+	return paramtypes.ParamSetPairs{
+		paramtypes.NewParamSetPair(KeyCreditClassFee, &p.CreditClassFee, validateCreditClassFee),
+		paramtypes.NewParamSetPair(KeyAllowedClassCreators, &p.AllowedClassCreators, validateAllowedClassCreators),
+		paramtypes.NewParamSetPair(KeyAllowlistEnabled, &p.AllowlistEnabled, validateAllowlistEnabled),
+		paramtypes.NewParamSetPair(KeyCreditTypes, &p.CreditTypes, validateCreditTypes),
+		paramtypes.NewParamSetPair(KeyBasketCreationFee, &p.BasketCreationFee, validateBasketCreationFee),
+	}
+}
+
+// Validate will run each param field's validate method
+func (p Params) Validate() error {
+	if err := validateCreditTypes(p.CreditTypes); err != nil {
+		return err
+	}
+
+	if err := validateAllowedClassCreators(p.AllowedClassCreators); err != nil {
+		return err
+	}
+
+	if err := validateAllowlistEnabled(p.AllowlistEnabled); err != nil {
+		return err
+	}
+
+	if err := validateCreditClassFee(p.CreditClassFee); err != nil {
+		return err
+	}
+
+	if err := validateBasketCreationFee(p.BasketCreationFee); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func validateCreditClassFee(i interface{}) error {
+	v, ok := i.(sdk.Coins)
+	if !ok {
+		return sdkerrors.ErrInvalidType.Wrapf("invalid parameter type: %T", i)
+	}
+
+	if err := v.Validate(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func validateAllowedClassCreators(i interface{}) error {
+	v, ok := i.([]string)
+	if !ok {
+		return sdkerrors.ErrInvalidType.Wrapf("invalid parameter type: %T", i)
+	}
+	for _, sAddr := range v {
+		_, err := sdk.AccAddressFromBech32(sAddr)
+		if err != nil {
+			return sdkerrors.ErrInvalidAddress.Wrapf("invalid creator address: %s", err.Error())
+		}
+	}
+	return nil
+}
+
+func validateAllowlistEnabled(i interface{}) error {
+	_, ok := i.(bool)
+	if !ok {
+		return sdkerrors.ErrInvalidType.Wrapf("invalid parameter type: %T", i)
+	}
+
+	return nil
+}
+
+func validateCreditTypes(i interface{}) error {
+	creditTypes, ok := i.([]*CreditType)
+	if !ok {
+		return sdkerrors.ErrInvalidType.Wrapf("invalid parameter type: %T", i)
+	}
+
+	// ensure no duplicate credit types or abbreviations and that all
+	// precisions conform to hardcoded PRECISION above
+	seenTypes := make(map[string]bool)
+	seenAbbrs := make(map[string]bool)
+	for _, creditType := range creditTypes {
+		// Validate name
+		T := core.NormalizeCreditTypeName(creditType.Name)
```

Comment:
Should this method be included in the migrations rather than using the method defined in core?

Reply:
make sense